### PR TITLE
Use caching ConsoleStream for both iOS and Android

### LINKED ISSUE #57304: Android + Console.WriteLine + format strings = possible unexpected extra newlines
### Description

Context: https://github.com/xamarin/xamarin-android/pull/6119
Context: https://github.com/xamarin/xamarin-android/pull/6119#issuecomment-896246633

.NET for Android "forwards" `Console.WriteLine()` invocations to `adb logcat`.  This is good and proper.

*However*, messages are written to `adb logcat` via [`__android_log_print`](https://developer.android.com/ndk/reference/group/logging#group___logging_1gac60e52c9711c53803ad9af639083afd5), and the `__android_log*` family of functions emits a newline after every invocation.

An odd "hiccup" is that if you use `Console.WriteLine()` with format strings to write a string which exceeds 11 characters (?!) in length, the value is written to `adb logcat` across two or more lines:

```csharp
using System;
using Android.App;

namespace android_hw
{
    [Activity(Label = "@string/app_name", MainLauncher = true)]
    public class MainActivity : Activity
    {
        protected override void OnCreate(Android.OS.Bundle savedInstanceState)
        {
            base.OnCreate(savedInstanceState);
            SetContentView(Resource.Layout.activity_main);

            Console.WriteLine ("{0}", new string ('x', 12));
        }
    }
}
```

…build & run…

```
% adb logcat > log.txt &
% ~/android-toolchain/dotnet/dotnet build /t:Install,StartAndroidActivity
```

…and `adb logcat` contains:

```
08-12 13:44:14.019 21107 21107 I DOTNET  : xxxxxxxxxxx
08-12 13:44:14.019 21107 21107 I DOTNET  : x
```

…which is, you know, *weird*.

There also doesn't appear to be any rhyme or reason for where the line wrap occurs, e.g. https://github.com/xamarin/xamarin-android/pull/6119#issuecomment-896242006

```csharp
try {
    throw new System.Exception("CRASH");
}
catch(System.Exception e)
{
    Console.WriteLine("# Unhandled Exception: e.ExceptionObject={0}", e);
}
```

results in:

```
08-10 16:02:49.498 16240 16267 I DOTNET  : # Unhandled Exception: e.ExceptionObject=System.Exce
08-10 16:02:49.498 16240 16267 I DOTNET  : ption: CRASH
08-10 16:02:49.498 16240 16267 I DOTNET  :    at Program.Main(String[] args) in /Users/thaysgrazia/runtime/src/mono/sample/Android/Program.cs:line 63
```

***Why does this matter?***

Sometimes you write tooling which reads `adb logcat` to see if a message was printed.  It is *extremely handy* if what you get is what you expect, and nobody expects "randomly inserted" newlines.

For example, [xamarin-android unit tests](https://github.com/xamarin/xamarin-android/blob/4fffaf8688db66b6353851bdfc5090ac34368de9/tests/MSBuildDeviceIntegration/Tests/InstallAndRunTests.cs#L64-L91).

For comparison, "traditional"/"legacy" Xamarin.Android only splits on `\n` boundaries before sending to `__android_log_print()`, so no "unexpected" newlines are present in `adb logcat` with legacy Xamarin.Android.

***Conjecture***:

What *might* be happening is `StreamWriter.WriteSpan()`: https://github.com/dotnet/runtime/blob/dbb2b53e0caeeb32888f83a5c6fe6db8525398de/src/libraries/System.Private.CoreLib/src/System/IO/StreamWriter.cs#L449-L466

`StreamWriter.Flush(bool, bool)` calls `LogcatStream.Write(ReadOnlySpan<byte>)`: https://github.com/dotnet/runtime/blob/dbb2b53e0caeeb32888f83a5c6fe6db8525398de/src/libraries/System.Console/src/System/ConsolePal.Android.cs#L16

`Interop.Logcat.AndroidLogPrint()`, meanwhile, is `__android_log_print()`: https://github.com/dotnet/runtime/blob/dbb2b53e0caeeb32888f83a5c6fe6db8525398de/src/libraries/Common/src/Interop/Android/Interop.Logcat.cs#L15

Every time `__android_log_print()` is invoked, a newline is written to `adb logcat`.

Thus, if the `while` loop in `StreamWriter.WriteSpan()` is executed more than once while trying to write the `# Unhandled Exception:` message, that *could* explain why `CRASH` is written on a separate line, and would explain what @thaystg described [elsewhere](https://github.com/xamarin/xamarin-android/pull/6119#issuecomment-896242006).

### Configuration

* Which version of .NET is the code running on? dotnet/installer/main@db9e71c7 6.0.100-rc.1.21408
* What OS and version, and what distro if applicable? Android 11 on a Pixel 5, plus various other Android targets
* What is the architecture (x64, x86, ARM, ARM64)? arm64, x86_64 (via emulator)
* Do you know whether it is specific to that configuration? It doesn't appear to be specific to Android version or hardware.  It instead appears to be specific to the Android-related .NET PAL & related code.

### Regression?

This is not a .NET regression, but it is a "regression" or "semantic change" from legacy Xamarin.Android.

### PR DESCRIPTION
**Pull request by steveisok**
Expands on #56713 by moving the caching implementation to a separate internal class and leaving only the interop calls in ConsolePal.iOS and ConsolePal.Android.

Fixes https://github.com/dotnet/runtime/issues/57304

### CODE REVIEW

**Review by filipnavara — APPROVED** [review]

File: src/libraries/System.Console/src/System/IO/CachedConsoleStream.cs [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+            finally
+            {
+                if (pooledBuffer != null)
+                {
+                    ArrayPool<char>.Shared.Return(pooledBuffer);
+                }
+            }
+        }
+
+        protected abstract unsafe void Print(ReadOnlySpan<char> line);
```
Comment (marek-safar) [inline src/libraries/System.Console/src/System/IO/CachedConsoleStream.cs:47]: It does not need to be unsafe here

File: src/libraries/System.Console/src/System/ConsolePal.Android.cs [hunk 2]
```diff
@@ -7,15 +7,13 @@
...
-    internal sealed unsafe class LogcatStream : ConsoleStream
+    internal sealed unsafe class LogcatStream : CachedConsoleStream
     {
-        public LogcatStream() : base(FileAccess.Write) {}
+        public LogcatStream(Encoding encoding) : base(encoding) {}
 
-        public override int Read(Span<byte> buffer) => throw Error.GetReadNotSupported();
-
-        public override unsafe void Write(ReadOnlySpan<byte> buffer)
+        protected override unsafe void Print(ReadOnlySpan<char> line)
```
Comment (marek-safar) [inline src/libraries/System.Console/src/System/ConsolePal.Android.cs:14]: Redundant `unsafe`

**Review by marek-safar — APPROVED** [review]

**Comment by steveisok** [comment]
Failures are unrelated.  Timing out most likely due to https://github.com/dotnet/core-eng/issues/14346

**Pull request merged**